src/gflownet/trainer.py

```diff
@@ -344,31 +378,66 @@ def run(self, logger=None):
         self._save_state(num_training_steps)
 
         num_final_gen_steps = self.cfg.num_final_gen_steps
+        final_info = {}
         if num_final_gen_steps:
             logger.info(f"Generating final {num_final_gen_steps} batches ...")
             for it, batch in zip(
-                range(num_training_steps, num_training_steps + num_final_gen_steps + 1),
+                range(num_training_steps + 1, num_training_steps + num_final_gen_steps + 1),
                 cycle(final_dl),
             ):
-                pass
-            logger.info("Final generation steps completed.")
+                if hasattr(batch, "extra_info"):
+                    for k, v in batch.extra_info.items():
+                        if k not in final_info:
+                            final_info[k] = []
+                        if hasattr(v, "item"):
+                            v = v.item()
+                        final_info[k].append(v)
+                if it % self.print_every == 0:
+                    logger.info(f"Generating mols {it - num_training_steps}/{num_final_gen_steps}")
+            final_info = {k: np.mean(v) for k, v in final_info.items()}
+
+            logger.info("Final generation steps completed - " + " ".join(f"{k}:{v:.2f}" for k, v in final_info.items()))
+            self.log(final_info, num_training_steps, "final")
+
+        # for pypy and other GC havers
+        del train_dl
+        del valid_dl
+        if self.cfg.num_final_gen_steps:
+            del final_dl
 
     def _save_state(self, it):
-        torch.save(
-            {
-                "models_state_dict": [self.model.state_dict()],
-                "cfg": self.cfg,
-                "step": it,
-            },
-            open(pathlib.Path(self.cfg.log_dir) / "model_state.pt", "wb"),
-        )
+        state = {
+            "models_state_dict": [self.model.state_dict()],
+            "cfg": self.cfg,
+            "step": it,
+        }
+        if self.sampling_model is not self.model:
+            state["sampling_model_state_dict"] = [self.sampling_model.state_dict()]
+        fn = pathlib.Path(self.cfg.log_dir) / "model_state.pt"
+        with open(fn, "wb") as fd:
+            torch.save(
+                state,
+                fd,
+            )
+        shutil.copy(fn, pathlib.Path(self.cfg.log_dir) / f"model_state_{it}.pt")
```

Comment:
Would be nice to have a flag here, saving every checkpoint can end up taking a lot of space so I wouldn't have it on by default.

Reply:
Sounds resonable.